File: 15 - Password Validation/Golang/main.go
package main

import (
    "bufio"
    "fmt"
    "os"
    "crypto/sha256"
    "encoding/hex"
)

func hash(s string) string {
    b := []byte(s)
    shad := sha256.Sum256(b)
    return hex.EncodeToString(shad[:32])
}

func validate(user string, pass string) bool {

    KNOWN := map[string]string {
        "piokozi" : "5e884898da28047151d0e56f8dc6292773603d0d6aabbdd62a11ef721d1542d8" ,
    }

    hashed := hash(pass)

    if _, exists := KNOWN[user]; !exists { // user does not exist
        return false
    } else if KNOWN[user] != hashed { // wrong password
        return false
    } else { // correct details
        return true
    }
}

func stdinput(o string) string {
    fmt.Printf("%s", o)
    scanner := bufio.NewScanner(os.Stdin)
    scanner.Scan()
    return scanner.Text()
}

func main() {
    username := stdinput("username: ")
    password := stdinput("password: ")

    successful := validate(username, password)

    if successful {
        fmt.Println("Welcome!")
    } else {
        fmt.Println("I don't know you.")
    }
}

File: 15 - Password Validation/Python/main.py
#!/usr/bin/env python

from hashlib import sha256

def validate(username, password):

    KNOWN = {
            # username : sha256
            "piokozi"  : "5e884898da28047151d0e56f8dc6292773603d0d6aabbdd62a11ef721d1542d8", # "password"
            }

    hashed = sha256(password.encode()).hexdigest()

    if username not in KNOWN.keys(): # the user does not exist
        return False
    elif hashed != KNOWN[username]: # the password is wrong
        return False
    else: # login is successful
        return True

if __name__ == "__main__":
    username = input("username: ")
    password = input("password: ")
    successful = validate(username, password)
    if successful:
        print("Welcome!")
    if not successful:
        print("I don't know you.")
